### ISSUE TITLE
Add Contingency Table Object

### ISSUE THREAD

**Issue by ebolyen** [biocore/scikit-bio#821]
`A user interacts with a contingency table to perform common ecological statistics including rarefaction, alpha, and beta diversity`.

**Comment by mortonjt** [comment]
Maybe I can start pitching in on this effort.  What are some of the critical portions of Biom that need to be ported over?

From what I can tell, the following portions should be ported
- [ ] internal scipy sparse matrix data type
- [ ] transform()
- [ ] reduce()
- [ ] filter()
- [ ] sort()
- [ ] ids()
- [ ] data()
- [ ] merge()

**Comment by mortonjt** [comment]
This may be breaking the ice.

Is there a particularly _strong_ reason why we can't include the BIOM dependency and import the BIOM table as a core data type?

It seems to me that BIOM already serves as a pretty darn good contingency table.  All it needs is an apply() function (like in pandas) and all of the ecological functions developed in skbio can be immediately applied to the BIOM tables.  Not only could it be a much cleaner solution, but it would bypass a lot of code duplication.

**Comment by wasade** [comment]
@mortonjt, there had been discussion on replacing the in-memory representation with a Pandas `DataFrame` but I do not know if that is the current plan. 

It might be easier to get the object integrated first, and replace the underlying type later.

**Comment by mortonjt** [comment]
Sounds like there is already a set plan on how to integrate this skbio.

I think Pandas DataFrames is certainly a sound approach.  But if any of the functionality from Biom is to be replicated with Pandas DataFrames, I would think that 3 separate data frames will be necessary (1 main dataframe and 2 separate metadata dataframes for observations and samples)

Once I get a good idea about how this should be implemented, I would be down to submit a PR.  Otherwise, I'm willing to review any upcoming PRs concerning this issue.

**Comment by ElDeveloper** [comment]
I agree, I think just moving the object as is at first would be a great
start, specially as other projects depend on those methods being
available, etc.

An idea that has already been suggested a few times is that the first
step should probably be providing an experimental method/function to
convert a BIOM table into a pandas DataFrame.

**Comment by ebolyen** [comment]
Sorry for not dropping in sooner.

@mortonjt We'd be really happy to have you help out with this endeavour, but there are a couple things that should have been mentioned as they have only really been discussed offline with @jairideout, @gregcaporaso, and myself. (It would be really cool to try out gitter.im this week and seeing how it works out, might prevent confusion as to our plans).

One of the very specific goals of this contingency table will be to decouple the file format from the object. `biom` != `ContingencyTable` and what we would like to see is all of the knowledge regarding the `biom-format` would be held in the io registry under a subpackage `skbio.io.biom` which will have 3 modules for each version of `biom`. So it might look like:
```
skbio.io
     - biom
           biom1_0.py
           biom2_0.py
           biom2_1.py
           __init__.py
```

We would then implement skbio sniffers for each variation of biom.

This endeavour will take a lot of effort, so our initial plan was to implement the core object first and focusing on what an API for that will look like and how it will fit with the rest of scikit-bio (this is something we are basically doing for every object before beta). With this in mind, the code from biom (unrelated to parsing) is less interesting to us than the functionality that the table API provides. 

What myself and @jairideout are going to be focusing on this week is not necessarily moving code around, but to try to get at what a contingency table means and how it should behave, the biom table provides some good inspiration, but we don't want to limit ourselves either.

**Comment by wasade** [comment]
I like this plan a lot. One thing to note, the majority of the biom code
base is the Table object (though parsing/formatting logic is partially
included). But completely agree that biom is not the object.

**Comment by wasade** [comment]
BTW, lemme know when you're going to get on gitter.im, would like to test
it out

**Comment by Jorge-C** [comment]
+1! I like that approach a lot.

**Comment by gregcaporaso** [comment]
Thanks everyone, I also like this approach a lot! I agree re: gitter.im. Ideally we'll have unlimited users, all discussion public, unlimited history. Depending on the cost, I can pay for the service.

**Comment by wasade** [comment]
So the gitter room now [exists](https://gitter.im/biocore/scikit-bio)
